src/modes/draw/ellipse.ts

```diff
@@ -0,0 +1,172 @@
+      });
+
+      if (this.flags.featureCreateAllowed) {
+        this.xSemiAxisLngLat = lngLat;
+        this.xSemiAxisPoint = this.gm.mapAdapter.project(this.circleCenterLngLat);
+      }
+    } else {
+      this.fireBeforeFeatureCreate({ geoJsonFeatures: [this.getFeatureGeoJson(lngLat)] });
+
+      if (this.flags.featureCreateAllowed) {
+        this.circleCenterLngLat = lngLat;
+        this.circleCenterPoint = this.gm.mapAdapter.project(this.circleCenterLngLat);
+        this.featureData = this.createFeature();
+
+        const markerData = this.getControlMarkerData();
+        if (this.featureData && markerData) {
+          this.fireStartEvent(this.featureData, markerData);
+        }
+      }
+    }
+
+    return { next: false };
+  }
+
+  onMouseMove(): MapHandlerReturnData {
+    if (this.circleCenterLngLat && this.gm.markerPointer.marker) {
+      const markerLngLat = this.gm.markerPointer.marker.getLngLat();
+
+      this.fireBeforeFeatureCreate({
+        geoJsonFeatures: [
+          this.xSemiAxisLngLat
+            ? this.getEllipseGeoJson(this.circleCenterLngLat, this.xSemiAxisLngLat, markerLngLat)
+            : this.getEllipseGeoJson(this.circleCenterLngLat, markerLngLat),
+        ],
+      });
+
+      if (this.flags.featureCreateAllowed) {
+        this.throttledMethods.updateFeatureGeoJson(markerLngLat);
+      }
+    }
+
+    if (!this.circleCenterLngLat) {
+      this.fireMarkerPointerUpdateEvent();
+    }
+
+    return { next: false };
+  }
+
+  updateFeatureGeoJson(eventLngLat: LngLat) {
+    if (!this.featureData || !this.circleCenterLngLat) {
+      return;
+    }
+
+    const featureGeoJson = this.xSemiAxisLngLat
+      ? this.getEllipseGeoJson(this.circleCenterLngLat, this.xSemiAxisLngLat, eventLngLat)
+      : this.getEllipseGeoJson(this.circleCenterLngLat, eventLngLat);
+    this.featureData.updateGeoJsonGeometry(featureGeoJson.geometry);
+
+    this.featureData.updateGeoJsonProperties({
+      shape: featureGeoJson.properties.shape,
+    });
+
+    const markerData = this.getControlMarkerData();
+    if (markerData) {
+      this.fireUpdateEvent(this.featureData, markerData);
+    }
+  }
+
+  saveEllipseFeature(eventLngLat: LngLat) {
+    if (!this.circleCenterLngLat || !this.xSemiAxisLngLat) {
+      return;
+    }
+
+    if (this.featureData) {
+      const lngLat = this.gm.markerPointer.marker?.getLngLat() || eventLngLat;
+      this.updateFeatureGeoJson(lngLat);
+
+      const { xSemiAxis, ySemiAxis, angle } = getEllipseParameters({
+        center: this.circleCenterLngLat,
+        xSemiAxisLngLat: this.xSemiAxisLngLat,
+        rimLngLat: lngLat,
+      });
+
+      this.featureData.setShapeProperty('center', this.circleCenterLngLat);
+      this.featureData.setShapeProperty('xSemiAxis', xSemiAxis);
+      this.featureData.setShapeProperty('ySemiAxis', ySemiAxis);
+      this.featureData.setShapeProperty('angle', angle);
+
+      if (this.isFeatureGeoJsonValid()) {
```

Comment:
Hi @zxwild,  
It's up to you to decide whether you want to homogenize with the `saveCircleFeature` method.  
At the time, it seemed a bit more complicated to clone with `shapeParameters` (when the featureData doesn't have a `_geoJson` property).

Here it only takes two lines.
But maybe I missed something in the logic, hence my comment.